File: src/ios/DogWalker/Presentation/Scenes/Auth/RegisterViewModel.swift
// Combine framework - Latest
import Combine

// Internal imports with relative paths
import "../../../Domain/UseCases/Auth/RegisterUseCase"
import "../../../Utilities/Logger"

/// RegisterViewModel: Manages the state and logic for the user registration screen
/// Requirements addressed:
/// - User Management (1.3 Scope/Core Features/User Management): Supports user registration and profile creation for dog owners and walkers
class RegisterViewModel: ObservableObject {
    // MARK: - Human Tasks
    /*
    1. Configure proper validation rules for password strength
    2. Set up error tracking service for registration failures
    3. Review and implement proper data sanitization rules
    4. Configure analytics tracking for registration events
    */
    
    // MARK: - Published Properties
    
    @Published var name: String?
    @Published var email: String?
    @Published var password: String?
    @Published var phoneNumber: String?
    @Published var role: String?
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    
    // MARK: - Private Properties
    
    private let registerUseCase: RegisterUseCase
    private var cancellables = Set<AnyCancellable>()
    
    // MARK: - Initialization
    
    /// Initializes the RegisterViewModel with required dependencies
    /// - Parameter registerUseCase: The use case handling user registration
    init(registerUseCase: RegisterUseCase) {
        self.registerUseCase = registerUseCase
        setupInitialState()
    }
    
    // MARK: - Private Methods
    
    /// Sets up the initial state of the ViewModel
    private func setupInitialState() {
        name = nil
        email = nil
        password = nil
        phoneNumber = nil
        role = nil
        isLoading = false
        errorMessage = nil
    }
    
    /// Validates the registration input fields
    /// - Parameters:
    ///   - name: User's full name
    ///   - email: User's email address
    ///   - password: User's password
    ///   - phoneNumber: User's phone number
    ///   - role: User's role (owner/walker)
    /// - Returns: Boolean indicating whether the input is valid
    private func validateInput(
        name: String,
        email: String,
        password: String,
        phoneNumber: String,
        role: String
    ) -> Bool {
        // Validate name
        guard !name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            errorMessage = "Name cannot be empty"
            return false
        }
        
        // Validate email format
        let emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
        let emailPredicate = NSPredicate(format: "SELF MATCHES %@", emailRegex)
        guard emailPredicate.evaluate(with: email) else {
            errorMessage = "Invalid email format"
            return false
        }
        
        // Validate password strength
        guard password.count >= 8,
              password.rangeOfCharacter(from: .uppercaseLetters) != nil,
              password.rangeOfCharacter(from: .lowercaseLetters) != nil,
              password.rangeOfCharacter(from: .decimalDigits) != nil else {
            errorMessage = "Password must be at least 8 characters and contain uppercase, lowercase, and numbers"
            return false
        }
        
        // Validate phone number format
        let phoneRegex = "^\\+?[1-9]\\d{1,14}$"
        let phonePredicate = NSPredicate(format: "SELF MATCHES %@", phoneRegex)
        guard phonePredicate.evaluate(with: phoneNumber) else {
            errorMessage = "Invalid phone number format"
            return false
        }
        
        // Validate role
        let validRoles = ["owner", "walker"]
        guard validRoles.contains(role.lowercased()) else {
            errorMessage = "Role must be either 'owner' or 'walker'"
            return false
        }
        
        return true
    }
    
    // MARK: - Public Methods
    
    /// Initiates the user registration process
    /// - Parameters:
    ///   - name: User's full name
    ///   - email: User's email address
    ///   - password: User's password
    ///   - phoneNumber: User's phone number
    ///   - role: User's role (owner/walker)
    func registerUser(
        name: String,
        email: String,
        password: String,
        phoneNumber: String,
        role: String
    ) {
        Logger.info("Starting registration process for user: \(email)")
        
        // Reset error message
        errorMessage = nil
        
        // Validate input
        guard validateInput(
            name: name,
            email: email,
            password: password,
            phoneNumber: phoneNumber,
            role: role
        ) else {
            Logger.error("Registration validation failed for user: \(email)")
            return
        }
        
        // Set loading state
        isLoading = true
        
        // Execute registration
        let result = registerUseCase.execute(
            name: name,
            email: email,
            password: password,
            phoneNumber: phoneNumber,
            role: role
        )
        
        // Handle result
        switch result {
        case .success(let user):
            Logger.info("Successfully registered user: \(user.email)")
            isLoading = false
            // Note: Further handling of successful registration (e.g., navigation)
            // should be implemented by the view layer
            
        case .failure(let error):
            Logger.error("Registration failed for user \(email): \(error.localizedDescription)")
            isLoading = false
            errorMessage = error.localizedDescription
        }
    }
}

File: src/android/app/src/main/java/com/dogwalker/app/presentation/screens/auth/RegisterViewModel.kt
package com.dogwalker.app.presentation.screens.auth

// androidx.lifecycle version: 2.6.0
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope

// kotlinx.coroutines version: 1.6.4
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

import com.dogwalker.app.domain.usecase.auth.RegisterUseCase
import com.dogwalker.app.domain.model.User

/**
 * Human Tasks:
 * 1. Verify error handling strategy aligns with application-wide error handling policy
 * 2. Configure proper logging for registration events
 * 3. Review input validation rules with product team
 * 4. Ensure proper error message localization is implemented
 */

/**
 * ViewModel that manages the state and logic for the RegisterScreen.
 *
 * Requirement addressed: User Management (1.3 Scope/Core Features/User Management)
 * Implements the presentation layer logic for user registration by managing form state
 * and coordinating with the RegisterUseCase.
 */
class RegisterViewModel(
    private val registerUseCase: RegisterUseCase
) : ViewModel() {

    companion object {
        const val REGISTER_SUCCESS = "Registration successful"
        const val REGISTER_FAILURE = "Registration failed"
    }

    private val _registrationState = MutableLiveData<Boolean>()
    val registrationState: LiveData<Boolean> = _registrationState

    private val _errorState = MutableLiveData<String>()
    val errorState: LiveData<String> = _errorState

    /**
     * Handles the user registration process by validating inputs and invoking the RegisterUseCase.
     *
     * @param name The full name of the user
     * @param email The email address for registration
     * @param password The password for the account
     *
     * Requirement addressed: User Management (1.3 Scope/Core Features/User Management)
     * Implements user registration by validating input parameters and coordinating
     * with the RegisterUseCase to create new user accounts.
     */
    fun register(name: String, email: String, password: String) {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                // Validate inputs before proceeding
                validateInputs(name, email, password)

                // Attempt registration through the use case
                registerUseCase.execute(
                    name = name,
                    email = email,
                    password = password
                )

                // Update success state
                _registrationState.postValue(true)
                _errorState.postValue(REGISTER_SUCCESS)
            } catch (e: IllegalArgumentException) {
                // Handle validation errors
                _registrationState.postValue(false)
                _errorState.postValue(e.message ?: REGISTER_FAILURE)
            } catch (e: Exception) {
                // Handle registration errors
                _registrationState.postValue(false)
                _errorState.postValue(e.message ?: REGISTER_FAILURE)
            }
        }
    }

    /**
     * Validates the registration input parameters.
     *
     * @param name The user's full name
     * @param email The user's email address
     * @param password The user's password
     * @throws IllegalArgumentException if any validation fails
     */
    private fun validateInputs(name: String, email: String, password: String) {
        require(name.isNotBlank()) { "Name cannot be blank" }
        require(email.isNotBlank()) { "Email cannot be blank" }
        require(password.isNotBlank()) { "Password cannot be blank" }

        require(name.length <= User.MAX_NAME_LENGTH) {
            "Name cannot exceed ${User.MAX_NAME_LENGTH} characters"
        }
        require(email.length <= User.MAX_EMAIL_LENGTH) {
            "Email cannot exceed ${User.MAX_EMAIL_LENGTH} characters"
        }
        require(email.matches(Regex("[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}"))) {
            "Invalid email format"
        }
        require(password.length >= 8) { "Password must be at least 8 characters long" }
    }
}